src/layers/shaderlib/fp64/cos-fp64.glsl

```diff
@@ -0,0 +1,143 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+#pragma glslify: sum_fp64 = require(./sum-fp64, ONE=ONE)
+#pragma glslify: sub_fp64 = require(./sub-fp64, ONE=ONE)
+#pragma glslify: mul_fp64 = require(./mul-fp64, ONE=ONE)
+#pragma glslify: div_fp64 = require(./div-fp64, ONE=ONE)
+#pragma glslify: sin_taylor_fp64 = require(./sin-taylor-fp64, ONE=ONE)
+#pragma glslify: cos_taylor_fp64 = require(./cos-taylor-fp64, ONE=ONE)
+#pragma glslify: sincos_taylor_fp64 = require(./sincos-taylor-fp64, ONE=ONE)
+#pragma glslify: nint_fp64 = require(./nint-fp64, ONE=ONE)
+
+const vec2 TWO_PI = vec2(6.2831854820251465, -1.7484556025237907e-7);
+const vec2 PI_2 = vec2(1.5707963705062866, -4.371139006309477e-8);
+const vec2 PI_16 = vec2(0.19634954631328583, -5.463923757886846e-9);
+const vec2 PI_16_2 = vec2(0.39269909262657166, -1.0927847515773692e-8);
+const vec2 PI_16_3 = vec2(0.5890486240386963, -1.4906100798128818e-9);
+
+const vec2 SIN_TABLE_0 = vec2(0.19509032368659973, -1.6704714833615242e-9);
+const vec2 SIN_TABLE_1 = vec2(0.3826834261417389, 6.22335089017767e-9);
+const vec2 SIN_TABLE_2 = vec2(0.5555702447891235, -1.1769521357507529e-8);
+const vec2 SIN_TABLE_3 = vec2(0.7071067690849304, 1.2101617041793133e-8);
+
+const vec2 COS_TABLE_0 = vec2(0.9807852506637573, 2.9739473106360492e-8);
+const vec2 COS_TABLE_1 = vec2(0.9238795042037964, 2.8307490351764386e-8);
+const vec2 COS_TABLE_2 = vec2(0.8314695954322815, 1.6870263741530778e-8);
+const vec2 COS_TABLE_3 = vec2(0.7071067690849304, 1.2101617152815436e-8);
+
+vec2 cos_fp64(vec2 a) {
+    if (a.x == 0.0 && a.y == 0.0) {
+        return vec2(1.0, 0.0);
+    }
+
+    // 2pi range reduction
+    vec2 z = nint_fp64(div_fp64(a, TWO_PI));
+    vec2 r = sub_fp64(a, mul_fp64(TWO_PI, z));
+
+    vec2 t;
+    float q = floor(r.x / PI_2.x + 0.5);
+    int j = int(q);
+
+    if (j < -2 || j > 2) {
+        return vec2(0.0 / 0.0, 0.0 / 0.0);
+    }
+
+    t = sub_fp64(r, mul_fp64(PI_2, vec2(q, 0.0)));
+
+    q = floor(t.x / PI_16.x + 0.5);
+    int k = int(q);
+
+    if (k == 0) {
+        if (j == 0) {
+            return cos_taylor_fp64(t);
+        } else if (j == 1) {
+            return -sin_taylor_fp64(t);
+        } else if (j == -1) {
+            return sin_taylor_fp64(t);
+        } else {
+            return -cos_taylor_fp64(t);
+        }
+    }
+
+        int abs_k = int(abs(float(k)));
+
+    // We just can't get PI/16 * 3.0 very accurately.
+    // so let's just store it
+    if (abs_k > 4) {
+        return vec2(0.0 / 0.0, 0.0 / 0.0);
+    } else if (k == 3) {
+        t = sub_fp64(t, PI_16_3);
+    } else if (k == -3) {
+        t = sum_fp64(t, PI_16_3);
+    } else {
+        t = sub_fp64(t, mul_fp64(PI_16, vec2(q, 0.0)));
+    }
+
+    vec2 u = vec2(0.0, 0.0);
+    vec2 v = vec2(0.0, 0.0);
+
+    if (abs_k == 1) {
```

Comment:
Just a thought... Would a switch statement be more appropriate? There's a few more places where it could be considered, but I guess it is borderline, with four cases.

Maybe keeping it this way keeps it closer to reference code...